compiler/rustc_parse/src/parser/path.rs

```diff
@@ -503,35 +559,141 @@ impl<'a> Parser<'a> {
-        let start = self.token.span;
         let arg = if self.check_lifetime() && self.look_ahead(1, |t| !t.is_like_plus()) {
             // Parse lifetime argument.
             GenericArg::Lifetime(self.expect_lifetime())
         } else if self.check_const_arg() {
             // Parse const argument.
-            let value = if let token::OpenDelim(token::Brace) = self.token.kind {
+            let expr = if let token::OpenDelim(token::Brace) = self.token.kind {
                 self.parse_block_expr(
                     None,
                     self.token.span,
                     BlockCheckMode::Default,
                     ast::AttrVec::new(),
                 )?
+            } else if self.token.is_ident() {
+                // FIXME(const_generics): to distinguish between idents for types and consts,
+                // we should introduce a GenericArg::Ident in the AST and distinguish when
+                // lowering to the HIR. For now, idents for const args are not permitted.
+                if self.token.is_bool_lit() {
+                    self.parse_literal_maybe_minus()?
+                } else {
+                    let span = self.token.span;
+                    let msg = "identifiers may currently not be used for const generics";
+                    self.struct_span_err(span, msg).emit();
+                    let block = self.mk_block_err(span);
+                    self.mk_expr(span, ast::ExprKind::Block(block, None), ast::AttrVec::new())
+                }
             } else {
-                self.handle_unambiguous_unbraced_const_arg()?
+                self.parse_literal_maybe_minus()?
             };
-            GenericArg::Const(AnonConst { id: ast::DUMMY_NODE_ID, value })
+            GenericArg::Const(AnonConst { id: ast::DUMMY_NODE_ID, value: expr })
         } else if self.check_type() {
             // Parse type argument.
-            match self.parse_ty() {
-                Ok(ty) => GenericArg::Type(ty),
-                Err(err) => {
-                    // Try to recover from possible `const` arg without braces.
-                    return self.recover_const_arg(start, err).map(Some);
-                }
-            }
+            GenericArg::Type(self.parse_ty()?)
         } else {
             return Ok(None);
         };
         Ok(Some(arg))
     }
+
+    fn get_assoc_type_with_generic_args(
+        &self,
+        gen_arg: Option<GenericArg>,
+        lo: Span,
+    ) -> PResult<'a, (Option<Ident>, Vec<GenericArg>)> {
+        if let Some(arg) = gen_arg {
+            match arg {
+                GenericArg::Type(t) => match &(*t).kind {
+                    TyKind::Path(_, path) => {
+                        if path.segments.len() == 1 {
+                            let path_seg = &path.segments[0];
+                            let ident = path_seg.ident;
+                            let gen_args = self.get_generic_args_from_path_segment(&path_seg)?;
```

Comment:
It's not a parser's job to reject `<A<X = Y> = B>` or `<Fn(X, Y) = B>` (especially the former one), so I think `get_generic_args_from_path_segment` can be removed.
These forms only being applicable to traits is a semantic property which is enforced during type checking, so parser can just pass this through.
(We may need this later if we ever have associated traits in the future.)

As a result, `AssocTyConstraint` needs to keep whole `GenericArgs` instead of `Vec<GenericArg>`.